core/src/main/java/com/linecorp/armeria/client/RedirectingClient.java

```diff
@@ -0,0 +1,299 @@
+import java.util.function.BiPredicate;
+
+import javax.annotation.Nullable;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.Multimap;
+
+import com.linecorp.armeria.common.HttpHeaderNames;
+import com.linecorp.armeria.common.HttpMethod;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.HttpRequestDuplicator;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.HttpStatusClass;
+import com.linecorp.armeria.common.RequestHeaders;
+import com.linecorp.armeria.common.RequestHeadersBuilder;
+import com.linecorp.armeria.common.ResponseHeaders;
+import com.linecorp.armeria.common.logging.RequestLogBuilder;
+import com.linecorp.armeria.common.logging.RequestLogProperty;
+import com.linecorp.armeria.common.stream.AbortedStreamException;
+import com.linecorp.armeria.internal.client.ClientUtil;
+
+final class RedirectingClient extends SimpleDecoratingHttpClient {
+
+    private static final Logger logger = LoggerFactory.getLogger(RedirectingClient.class);
+
+    private final RedirectConfig redirectConfig;
+    private final boolean withBaseUri;
+
+    RedirectingClient(HttpClient delegate, RedirectConfig redirectConfig, boolean withBaseUri) {
+        super(delegate);
+        this.redirectConfig = redirectConfig;
+        this.withBaseUri = withBaseUri;
+    }
+
+    @Override
+    public HttpResponse execute(ClientRequestContext ctx, HttpRequest req) throws Exception {
+        final CompletableFuture<HttpResponse> responseFuture = new CompletableFuture<>();
+        final HttpResponse res = HttpResponse.from(responseFuture, ctx.eventLoop());
+        final HttpRequestDuplicator reqDuplicator = req.toDuplicator(ctx.eventLoop().withoutContext(), 0);
+        final RedirectContext redirectCtx = new RedirectContext(req, res, responseFuture);
+        execute0(ctx, redirectCtx, reqDuplicator, true);
+        return res;
+    }
+
+    private void execute0(ClientRequestContext ctx, RedirectContext redirectCtx,
+                          HttpRequestDuplicator reqDuplicator, boolean initialAttempt) {
+        final CompletableFuture<Void> originalReqWhenComplete = redirectCtx.request().whenComplete();
+        final CompletableFuture<HttpResponse> responseFuture = redirectCtx.responseFuture();
+        if (originalReqWhenComplete.isCompletedExceptionally()) {
+            originalReqWhenComplete.exceptionally(cause -> {
+                handleException(ctx, reqDuplicator, responseFuture, cause, initialAttempt);
+                return null;
+            });
+            return;
+        }
+
+        if (redirectCtx.responseWhenComplete().isDone()) {
+            redirectCtx.responseWhenComplete().handle((result, cause) -> {
+                final Throwable abortCause = firstNonNull(cause, AbortedStreamException.get());
```

Comment:
> Could use if ... else to avoid an additional allocation of AbortedStreamException?

That's a good point. 👍 Honestly, I just copied this from `RetryingClient`. 😅 Let me change both.